### PR TITLE
Baggage propagator encoding/decoding fix

### LINKED ISSUE #1607: Baggage value certain characters will crash application or encode chars
# Bug Report

```xml
      <PackageReference Include="OpenTelemetry.Api" Version="1.0.0-rc1.1" />
      <PackageReference Include="OpenTelemetry.Exporter.Jaeger" Version="1.0.0-rc1.1" />
      <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.0.0-rc1.1" />
      <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc1.1" />
      <PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.0.0-rc1.1" />
```

* Runtime version: `netcoreapp3.1`

## Symptom

When adding `Baggage` to an `Activity` **using certain chars will encode them or crash the receiving application.**
The baggage value is not correctly received in child-service.

| Char | Result                              |
| ---- | ----------------------------------- |
| `\`  | `%2B`                               |
| `@`  | `%2540`                             |
| `,`  | `%252C`                             |
| `&`  | `%2526`                             |
| `(`  | Will crash the ASP.NET Core Service |
(only some samples)

There is no mentioning of a limited allowed charset in [specification/overview.md#baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/overview.md#baggage) ?

**What is the expected behavior?**

Original baggage value should be preserved exactly.

**What is the actual behavior?**

Some characters are only encoded.

**Using round brackets will crash the ASP.NET Core WEB API Service.**

Using `.AddBaggage("demo.foobar5", "!x_x,x-x&x(x");`:
```
[2020-11-22 15:23:48Z ERR] Connection id "0HM4EVAE30FTQ", Request id "0HM4EVAE30FTQ:00000001": 
An unhandled exception was thrown by the application. System.FormatException: The format of
value '!x_x%2Cx-x%26x(x' is invalid.
   at System.Net.Http.Headers.NameValueHeaderValue.CheckValueFormat(String value)
   at System.Net.Http.DiagnosticsHandler.InjectHeaders(Activity currentActivity, HttpRequestMessage request)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at FrontendService.Controllers.WeatherFrontendController.GetWeatherForecast() in D:xxxxxxxxxxxxx:line 109

```

## Reproduce

```csharp
Activity.Current?
	.AddBaggage("demo.foobar1", "demouser")
	.AddBaggage("demo.foobar2", "demo user")
	.AddBaggage("demo.foobar3", @"domain\demouser")
	.AddBaggage("demo.foobar4", "user@example.com")
        .AddBaggage("demo.foobar5", "!x_x,x-x&x(x"); // Will cause exception in receiving service
```

In child-service:
```csharp
if (Activity.Current?.Baggage != null)
{
	foreach (var bag in Activity.Current.Baggage)
	{
		_logger.LogInformation("Found Baggage {BaggageKey}={BaggageValue}", bag.Key, bag.Value);

		// Use baggage data in tag which will be stored in jaeger
		Activity.Current.SetTag($"ax.debug.{bag.Key}", bag.Value);
	}
}
```

### Log output of receiving service
```
Found Baggage demo.foobar4=user%2540example.com 
Found Baggage demo.foobar3=domain%255Cdemouser 
Found Baggage demo.foobar2=demo%2Buser 
Found Baggage demo.foobar1=demouser 
```

### Resulting display in jaeger `1.21.0`:

grafik

### PR DESCRIPTION
**Pull request by kipwoker**
Fixes #1607.

## Changes
- Add `WebUtility.UrlDecode` to the `Extract` method. During injection, propagator encoded values that should be decoded during extraction
- Remove `NameValueHeaderValue` usage, since [it was made for encoding/decoding/parsing](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2012#discussion_r633695125), but it doesn't align with the expectations. It doesn't parse input containing special symbols correctly
- Add tests for special character cases

### CODE REVIEW

**Comment by kipwoker** [comment]
Can't reproduce the crash for `Linux / build-test (netcoreapp3.1)` check.
It says `BasicTests.AddAspNetCoreInstrumentation_BadArgs` crashed. I can't figure out how could I break it.
Could you please help me to with it? :)

**Comment by kipwoker** [comment]
Strange thing. The first time Linux build was failed. And after merging master into the current branch it's green, but now Windows tests for netcore 3.1 fail. I also tried to reproduce this on my local machine, but everything is green.

@cijothomas could you take a look at this, please?

**Comment by cijothomas** [comment]
> Strange thing. The first time Linux build was failed. And after merging master into the current branch it's green, but now Windows tests for netcore 3.1 fail. I also tried to reproduce this on my local machine, but everything is green.
> 
> @cijothomas could you take a look at this, please?

https://github.com/open-telemetry/opentelemetry-dotnet/pull/2001#issuecomment-828195219 your current failure was this. I'll retrigger build. (or you can do so by simply closing, then opening this PR)

**Comment by kipwoker** [comment]
> [#2001 (comment)](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2001#issuecomment-828195219) your current failure was this. I'll retrigger build. (or you can do so by simply closing, then opening this PR)

Thank you. It helped. :)

**Comment by kipwoker** [comment]
@cijothomas again failed after merge
```csharp
Passed OpenTelemetry.Exporter.Zipkin.Tests.ZipkinExporterTests.IntegrationTest(useShortTraceIds: False, useTestResource: False, isRootSpan: False, statusCode: Error, statusDescription: null, addErrorTag: False) [1 ms]
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.InvalidOperationException: Please call the Start() method before calling this method.
   at System.Net.HttpListener.BeginGetContext(AsyncCallback callback, Object state)
   at System.Net.HttpListener.<>c.<GetContextAsync>b__44_0(AsyncCallback callback, Object state)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(Func`3 beginMethod, Func`2 endFunction, Action`1 endAction, Object state, TaskCreationOptions creationOptions)
   at System.Threading.Tasks.TaskFactory.FromAsync[TResult](Func`3 beginMethod, Func`2 endMethod, Object state)
   at System.Net.HttpListener.GetContextAsync()
   at OpenTelemetry.Tests.TestHttpServer.RunningServer.<>c__DisplayClass3_0.<<-ctor>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by xiang17** [comment]
> * `WebUtility.UrlEncode` doesn't encode round braces. That's why there was an exception in the `Extract` method during `NameValueHeaderValue.TryParse` execution. Replaced with `Uri.EscapeDataString` which much safer. It doesn't skip any character and behaves itself predictable (e.g. `WebUtility.UrlEncode` converts whitespace into `+` sign instead of `%20`)

The [document](https://docs.microsoft.com/en-us/dotnet/api/system.net.webutility.urlencode?view=net-5.0) for `WebUtility.UrlEncode` says it encodes "`(` and `)` (opening and closing parentheses)". Why was it causing this bug?

**Comment by kipwoker** [comment]
> > * `WebUtility.UrlEncode` doesn't encode round braces. That's why there was an exception in the `Extract` method during `NameValueHeaderValue.TryParse` execution. Replaced with `Uri.EscapeDataString` which much safer. It doesn't skip any character and behaves itself predictable (e.g. `WebUtility.UrlEncode` converts whitespace into `+` sign instead of `%20`)
> 
> The [document](https://docs.microsoft.com/en-us/dotnet/api/system.net.webutility.urlencode?view=net-5.0) for `WebUtility.UrlEncode` says it encodes "`(` and `)` (opening and closing parentheses)". Why was it causing this bug?

I see the different :)

> URL encoding replaces all character codes **except** for letters, numbers, and the following punctuation characters:

**Comment by kipwoker** [comment]
@cijothomas @alanwest @reyang 
Could you please take a look at this PR? :)

File: test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs [hunk 1]
```diff
@@ -159,5 +185,23 @@ public void ValidateBaggageInjection()
...
+                new Baggage(new Dictionary<string, string>
+                {
+                    { "key 1", "value 1" },
+                    { "key2", "!x_x,x-x&x(x\");:" },
+                }));
+
+            this.baggage.Inject(propagationContext, carrier, Setter);
+
+            Assert.Single(carrier);
+            Assert.Equal("key%201=value%201,key2=%21x_x%2Cx-x%26x%28x%22%29%3B%3A", carrier[BaggagePropagator.BaggageHeaderName]);
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs:204]: I think `!` can be optimized by just using one single character `!` instead of `%21` here.
Here goes the spec https://github.com/w3c/baggage/blob/master/baggage/HTTP_HEADER_FORMAT.md#definition.

Comment (reyang) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs:204]: The same for `_` vs. `%26`, `&` vs. `%28` and `(` vs. `%29`.

Comment (reyang) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs:204]: I think we want to have the implementation which generates the shortest representation given the W3C Baggage does put limitation on the size.

**Review by reyang — DISMISSED** [review]
Please refer to my comment [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2012#discussion_r629518477).

File: test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs (hunk 1 above)
Comment (kipwoker) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs:204]: This approach produces an exception. You can have a look at issue #1607 /
It looks like:
```
[2020-11-22 15:23:48Z ERR] Connection id "0HM4EVAE30FTQ", Request id "0HM4EVAE30FTQ:00000001": 
An unhandled exception was thrown by the application. System.FormatException: The format of
value '!x_x%2Cx-x%26x(x' is invalid.
   at System.Net.Http.Headers.NameValueHeaderValue.CheckValueFormat(String value)
   at System.Net.Http.DiagnosticsHandler.InjectHeaders(Activity currentActivity, HttpRequestMessage request)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at FrontendService.Controllers.WeatherFrontendController.GetWeatherForecast() in D:xxxxxxxxxxxxx:line 109
```

`System.Net.Http.Headers.NameValueHeaderValue.CheckValueFormat` throws an exception for such symbols. That's why I've made it safer. Another approach — to replace this check with something else. But it could be done by another PR.

The main purpose of this PR — make this code workable without serious performance degradation. At the moment the user can't use all of these symbols at all.

Comment (kipwoker) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs:204]: I think I got what you want. It's a good point. :)

Let me rephrase your sentence to ensure I got you right.
Am I right that you want the following flow:
1. We have raw baggage that was added like `!x_x,x-x&x(x`
2. Then we convert it with `WebUtility.UrlEncode` so the value will be transformed into `!x_x%2Cx-x%26x(x`
3. Then we send it to the recipient app
4. Then the recipient app decodes it with `WebUtility.UrlDecode` and got the raw value
5. But at this moment we have to call `NameValueHeaderValue.TryParse` and for this purpose, we encode the value again with `Uri.EscapeDataString` and then use `Uri.UnescapeDataString` after parsing

Is this correct? :)

Comment (kipwoker) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs]: I've made the changes. Please take a look :)

**Comment by kipwoker** [comment]
Can't understand why CodeCov failed on `ZPagesActivityTracker`. I don't think I could affect it by this PR.

Does anybody know what's wrong? :)

File: test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs (hunk 1 above)
Comment (reyang) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs]: The flow looks reasonable, there are room to improve the perf and memory use (e.g. reduce heap allocation) though (which I guess is not a blocker for this PR).

**Comment by reyang** [comment]
> Can't understand why CodeCov failed on `ZPagesActivityTracker`. I don't think I could affect it by this PR.
> 
> Does anybody know what's wrong? :)

CodeCov has known issues of being unstable, the coverage rate is an approximation rather than an accurate measurement.

File: test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs (hunk 1 above)
Comment (kipwoker) [reply test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs]: Agree, the performance could be improved in the separate PR.

**Comment by kipwoker** [comment]
> CodeCov has known issues of being unstable, the coverage rate is an approximation rather than an accurate measurement.

So, what should I do in this situation? I aimed to make this PR merged. :)

**Comment by cijothomas** [comment]
> > CodeCov has known issues of being unstable, the coverage rate is an approximation rather than an accurate measurement.
> 
> So, what should I do in this situation? I aimed to make this PR merged. :)

Code coverage is not a *Required* check for PR.

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs [hunk 2]
```diff
@@ -143,7 +143,10 @@ internal static bool TryExtractBaggage(string[] baggageCollection, out Dictionar
                         break;
                     }
 
-                    if (NameValueHeaderValue.TryParse(pair, out NameValueHeaderValue baggageItem))
+                    var decodedPair = WebUtility.UrlDecode(pair);
+                    var escapedPair = string.Join("=", decodedPair.Split('=').Select(Uri.EscapeDataString));
```
[the remark this answers is not in the document]
Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:147]: Well, we can check `pair` variable for `%` sign, but it may contain '(' or '!' (any symbol from Symptoms section #1607). If we will not escape them `NameValueHeaderValue.TryParse` will return false and we will lose the baggage. As I can see you are the author of this code. Could you please explain what's wrong will happen if we will remove `NameValueHeaderValue.TryParse` call and just split the `pair` by `=` sign?

[the remark this answers is not in the document]
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:147]: @kipwoker The `NameValueHeaderValue` stuff came from [Microsoft.AspNet.TelemetryCorrelation](https://github.com/aspnet/Microsoft.AspNet.TelemetryCorrelation/blob/b914ebf89d4db4568577af1db1e270f522b05f92/src/Microsoft.AspNet.TelemetryCorrelation/ActivityExtensions.cs#L115). The idea was it would take care of these encoding/decoding/parsing details for us. If it isn't working well, there's no reason to hold onto it.

[the remark this answers is not in the document]
Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs]: I wrote some tests. Looks like it works well without `NameValueHeaderValue` even if the user escaped their value instead of encoding. Now it looks much simpler to me, also it should work faster and consume less memory. What do you think?

**Comment by kipwoker** [comment]
@reyang @CodeBlanch please take a look at the new implementation. :)

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs [hunk 3]
```diff
@@ -143,20 +143,31 @@ internal static bool TryExtractBaggage(string[] baggageCollection, out Dictionar
...
+                    if (pair.IndexOf('=') < 0)
                     {
-                        if (string.IsNullOrEmpty(baggageItem.Name) || string.IsNullOrEmpty(baggageItem.Value))
-                        {
-                            continue;
-                        }
+                        continue;
+                    }
+
+                    var parts = pair.Split(new[] { '=' }, 2);
```
Comment (reyang) [inline src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:151]: Could we avoid "new array" each time? I wonder if `pair.Split('=', 2)` would just work?

Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:151]: You cannot pass the only one char in this particular overload. But it's a good point, thank you. I will put this array to the static field.

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs [hunk 4]
```diff
@@ -143,20 +143,31 @@ internal static bool TryExtractBaggage(string[] baggageCollection, out Dictionar
...
+                    if (parts.Length != 2)
+                    {
+                        continue;
+                    }
 
-                        if (baggageDictionary == null)
-                        {
-                            baggageDictionary = new Dictionary<string, string>();
-                        }
+                    var key = WebUtility.UrlDecode(parts[0]);
```
Comment (reyang) [inline src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:157]: Ideally (for best perf) we probably want to use [UrlDecode with offset](https://docs.microsoft.com/en-us/dotnet/api/system.web.httputility.urldecode?view=net-5.0#System_Web_HttpUtility_UrlDecode_System_Byte___System_Int32_System_Int32_System_Text_Encoding_). Not a blocker and can be a separate PR.

**Review by reyang — APPROVED** [review]
LGTM.

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs (hunk 4 above)
Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:157]: Could you please give me a doc to read why it's better for performance?

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs (hunk 3 above)
Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:151]: Also
```
foreach (var pair in item.Split(','))
```
This one also makes an array under the hood since the signature is `Split(params char[] ...)`

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs (hunk 4 above)
Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:160]: Also, if the UrlDecode is better for perf, is it means also that UrlEncode will be better during the injection process?

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs (hunk 3 above)
Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs]: > static sounds fine to me, curious if there is a better way (I'm not C# expert but gut was telling me there got to be a way 😄)

Yeah, `static` means it will be only one object to the whole application. Otherwise, the field will be created for each `BaggagePropagator` object. :)

File: src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs (hunk 4 above)
Comment (reyang) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:160]: I don't know a doc, but here goes the idea:

Instead of trying to split the string into smaller strings (which makes GC busy), we can potentially use an integer to store the index of `=` during the loop, and use the in-place version of UrlDecode, so that we avoid creating many intermediate strings.

This is the current code:

```csharp
foreach (var pair in item.split(',')) // this creates an enumerator, an array and several strings
{
  var parts = pair.split('=', ...); // this creates one array and two strings, for each loop
  decode(parts[0]); // this creates one result string
  decode(parts[1]); // this creates one result string
}
```

What we can potentially do:

```csharp
while (i < item.Length) // we only need few integers on the stack, no heap allocation needed for any intermediate value
{
  check if we got `=`, if yes, decode the key and remember the position
  check if we got the splitter `,`, if yes, decode the value with decode(item, startPosition, stopPosition)
}
```

Comment (kipwoker) [reply src/OpenTelemetry.Api/Context/Propagation/BaggagePropagator.cs:160]: Good point. I've got the idea. As you said before it can be a separate PR.

**Review by CodeBlanch — APPROVED** [review]
Thanks @kipwoker
